dm/syncer/ddl.go

```diff
@@ -245,18 +249,43 @@ func adjustCollation(tctx *tcontext.Context, ddlInfo *ddlInfo, statusVars []byte
 				tctx.L().Warn("not found charset default collation.", zap.String("originSQL", ddlInfo.originDDL), zap.String("charset", strings.ToLower(justCharset)))
 				return
 			}
-			tctx.L().Info("detect create database risk which use explicit charset and implicit collation, we will add collation by SHOW CHARACTER SET", zap.String("originSQL", ddlInfo.originDDL), zap.String("collation", collation))
+			tctx.L().Info("detect create database risk which use explicit charset and implicit collation, we will add collation by INFORMATION_SCHEMA.COLLATIONS", zap.String("originSQL", ddlInfo.originDDL), zap.String("collation", collation))
 		} else {
 			// has no charset and collation
 			// add collation by server collation from binlog statusVars
-			collation, _ = event.GetServerCollationByStatusVars(statusVars)
+			collation, _ = event.GetServerCollationByStatusVars(statusVars, idAndCollationMap)
 			// add collation
 			tctx.L().Info("detect create database risk which use implicit charset and collation, we will add collation by binlog status_vars", zap.String("originSQL", ddlInfo.originDDL), zap.String("collation", collation))
 		}
 		createStmt.Options = append(createStmt.Options, &ast.DatabaseOption{Tp: ast.DatabaseOptionCollate, Value: collation})
 	}
 }
 
+// adjustColumnsCollation adds column's collation.
+func adjustColumnsCollation(tctx *tcontext.Context, createStmt *ast.CreateTableStmt, charsetAndDefaultCollationMap map[string]string) {
```

Comment:
I'm not sure why there are two setting related to collation in `col.Options` and `col.Tp` 😵‍💫

Reply:
yes, it is strange. there is no `character set` in `col.Options`. so i just get charset from `col.Tp`, i will test the detail about them.

Reply:
```
mysql> create table t (id varchar(32) character set utf8mb4 collate utf8mb4_bin primary key collate utf8mb4_general_ci);
Query OK, 0 rows affected (0.10 sec)
```
But in my test, the `show create table` will only contains one of them so check both of them is ok here.

Reply:
ok, i will test.

Reply:
After some test, `collate` is not in `col.Tp` but only in `col.Options`. there will be two `col.Options` about `collate` when sql is `create table t (id varchar(32) character set utf8mb4 collate utf8mb4_bin primary key collate utf8mb4_general_ci)`. But when we set  `collate`  value to `col.Tp`, it will contains `collate`  info when rewrite sql by syntaxtree. 
By the way, the last `collate` will be effective when there are two `collate`. this is both in mysql 5.7.22 or tidb, but it will be an `ERROR 1064 (42000): Multiple COLLATE` in mysql 8.0.21.
So it seems ok in current test case but i think  we can also both check `col.Tp` and `col.Options` for safety.